content/blogposts/2024/anatomy-of-a-coding-assistant.md

```diff
@@ -0,0 +1,157 @@
+
+So, by fetching the right parts of your codebase and giving them to your AI coding assistant, the underlying AI model is able to come up with much better answers.
+
+![5_cody](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/5_cody.png)
+
+Over the past year or so, we’ve learned a lot about what works for fetching the right context for an AI coding assistant. In this post, I’m going to share some of those learnings and uncover how it works under the hood.
+
+---
+
+## An AI coding assistant - a product of products
+
+The reality is, if you look at a coding assistant like [Sourcegraph Cody](https://sourcegraph.com/cody), it’s not a single product—it’s many products put together. You have autocomplete, which makes suggestions as you code that keep you in flow. Then there’s chat, which allows you to ask questions of and generate code that fits in your codebase. Cody also has a feature that automates quick inline code changes with natural language, and another one that generates unit tests.
+
+![6_edits](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/6_edits.png)
+
+Each of these features works very differently from one another. The mental model of the user and the requirements are different for each. For autocomplete, speed is key. You want it to be very fast and accurate enough so that it doesn’t interrupt the programmer’s flow. On the other hand, for chat, accuracy and the quality of the responses may be more important.
+
+![7_tradeoff](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/7_tradeoff.png)
+
+When generating tests, there may be additional requirements. For example, if you’re looking at a file and want to add tests to it, you need to quickly detect if a test file for that file already exists. So, when we say “coding assistant,” it might sound simple, but it’s actually a product of products—multiple products put together to create a comprehensive experience.
+
+In this post, we're going to take a look at how context fetching works for each particular feature - chat, autocomplete, generating tests, and editing code.
+
+---
+
+## How context fetching works for chat
+
+![8_chat](https://storage.googleapis.com/sourcegraph-assets/blog/anatomy/8_chat.png)
+
+Chat allows you to ask questions about your codebase and generate code that's tailored to work within your codebase. For this, we want a few things:
+
+1. **Conversation history:** In a given chat session, we record previous messages as they may contain relevant context for the user's next request.
+2. **Code search:** We fetch the most relevant code snippets related to the user's query from the codebase using search, similar to how a human dev might search for these code snippets.
+3. **User control:** Users should have the ability to mention specific files and provide those directly to the model, and also include the option to reference external sources like Slack threads or Notion documents to enrich the context further.
```

Comment:
```suggestion
3. **User choice:** Users should have the ability to mention specific files and provide those directly to the model, and also include the option to reference external sources like Slack threads or Notion documents to enrich the context further.
```

maybe "choice" instead of "control"?

Reply:
Hmm here, I think "control" fits better. "Choice" makes me think of choosing an LLM model - so maybe more like choosing one thing out of several options.

Reply:
Makes sense. :)